recipes/RoutingPushReactHooks/src/Main.purs

```diff
@@ -0,0 +1,202 @@
+  React.component "PostIndex" \_ ->
+    pure do
+      R.ul_
+        ( Array.range 1 10
+            <#> \n ->
+                R.li_
+                  [ link
+                      { to: "/posts/" <> show n
+                      , children:
+                          [ R.text ("Post " <> show n) ]
+                      }
+                  ]
+        )
+
+mkPost :: Component Int
+mkPost = do
+  link <- mkLink
+  React.component "Post" \n ->
+    pure do
+      React.Basic.fragment
+        [ R.h1_ [ R.text ("Post " <> show n) ]
+        , R.p_
+            [ link
+                { to: "/posts/" <> show n <> "/edit"
+                , children: [ R.text "Click here" ]
+                }
+            , R.text " to edit this post"
+            ]
+        ]
+
+mkPostEdit :: Component Int
+mkPostEdit =
+  React.component "PostEdit" \n ->
+    pure (R.h1_ [ R.text ("Edit post " <> show n) ])
+
+data AppRoute
+  = PostIndex
+  | Post Int
+  | PostEdit Int
+  | Home
+
+appRoute :: Match (Maybe AppRoute)
+appRoute =
+  Foldable.oneOf
+    [ Just <$> postRoute
+    , Just <$> (Match.root *> pure Home <* Match.end)
+    , pure Nothing
+    ]
+  where
+  postRoute =
+    Match.root *> Match.lit "posts"
+      *> Foldable.oneOf
+          [ PostEdit <$> Match.int <* Match.lit "edit"
+          , Post <$> Match.int
+          , pure PostIndex
+          ]
+      <* Match.end
+
+type RouterContextValue
+  = { route :: Maybe AppRoute
+    , nav :: PushStateInterface
+    }
+
+-- | Note that we actually want a `RouterContextValue` where the context is
+-- | being consumed, not a `Maybe RouterContextValue`, but `createContext`
+-- | requires an "initial" value to use as a fallback in the case that the
+-- | context is used with no context provider. One solution would be to
+-- | construct a "dummy" value of type `RouterContextValue` that could work as a
+-- | sensible default. Another solution is to consider the use of the context
+-- | where it's not provided as *unintended behavior*, as described in this
+-- | article (in JS):
+-- | https://kentcdodds.com/blog/how-to-use-react-context-effectively.
+-- | *tl;dr* -- In JavaScript, the approach is to pass `undefined` or `null` as
+-- | the initial value and then instead of consuming the context directly at the
+-- | component level via `useContext`, to implement a custom hook that wraps
+-- | `useContext` and throws an error if the context is used where it's not
+-- | provided (signalling that this is not a use case we want to support). We've
+-- | done similar, by wrapping our context value in `Maybe` and using `Nothing`
+-- | as the case that we pattern-match on to trigger the error.
+routerContext :: ReactContext (Maybe RouterContextValue)
+routerContext =
+  -- | We are using `unsafePerformEffect` so that we can have a "global"
+  -- | instance of this context that we can use directly in our
+  -- | `useRouterContext` hook.
+  Effect.Unsafe.unsafePerformEffect do
+    React.createContext Nothing
```

Comment:
Now I'm not sure if any of this is the recommended approach, after reading through this: https://github.com/spicydonuts/purescript-react-basic-hooks/issues/41 😅 Will have to think on how best to refactor this.

Reply:
I've refactored to avoid `unsafePerformEffect`